src/main/java/org/terasology/metalrenegades/ai/actions/NearbyTargetInsideMarketAction.java

```diff
@@ -0,0 +1,90 @@
+import java.util.Random;
+import org.terasology.metalrenegades.ai.component.HomeComponent;
+import org.terasology.navgraph.WalkableBlock;
+
+/**
+ * Sets a marketkeeper character's {@link MinionMoveComponent} target to a random nearby block inside the area of
+ * their market building.
+ */
+@BehaviorAction(name = "set_target_inside_market")
+public class NearbyTargetInsideMarketAction extends BaseAction {
+
+    private static final Logger logger = LoggerFactory.getLogger(NearbyTargetInsideMarketAction.class);
+
+    private static final int RANDOM_BLOCK_ITERATIONS = 10;
+    private Random random = new Random();
+
+    private Rect2i marketRegion;
+
+    private int moveProbability = 100;
+
+    @Override
+    public void construct(Actor actor) {
+        HomeComponent homeComp = actor.getComponent(HomeComponent.class);
+        EntityRef buildingEntity = homeComp.building;
+
+        DynParcelRefComponent dynParcelRefComponent = buildingEntity.getComponent(DynParcelRefComponent.class);
+        DynParcel dynParcel = dynParcelRefComponent.dynParcel;
+
+        marketRegion = dynParcel.getShape();
+    }
+
+    @Override
+    public BehaviorState modify(Actor actor, BehaviorState result) {
+        if (random.nextInt(100) > (99 - moveProbability)) {
+            MinionMoveComponent moveComponent = actor.getComponent(MinionMoveComponent.class);
+            if (moveComponent.currentBlock != null) {
+                WalkableBlock target = findRandomNearbyBlockInMarket(moveComponent.currentBlock);
+                moveComponent.target = target.getBlockPosition().toVector3f();
+                actor.save(moveComponent);
+            } else {
+                return BehaviorState.FAILURE;
+            }
+        }
+        return BehaviorState.SUCCESS;
+    }
+
+    /**
+     * Finds a block close to the character, with the condition that this block must be inside the market parcel.
+     *
+     * @param startBlock The block that this characer is currently standing on.
+     * @return A random nearby block nearby inside the market area.
+     */
+    private WalkableBlock findRandomNearbyBlockInMarket(WalkableBlock startBlock) {
+        WalkableBlock currentBlock = startBlock;
+        for (int i = 0; i < random.nextInt(10) + 3; i++) {
+            WalkableBlock[] neighbors = currentBlock.neighbors;
+            List<WalkableBlock> existingNeighbors = Lists.newArrayList();
+
+            for (WalkableBlock neighbor : neighbors) {
+                if (neighbor != null && marketRegion.contains(neighbor.x(), neighbor.z())) {
+                    existingNeighbors.add(neighbor);
+                }
+            }
+            if (existingNeighbors.size() > 0) {
+                currentBlock = existingNeighbors.get(random.nextInt(existingNeighbors.size()));
+            }
+        }
+        logger.debug(String.format("Looking for a block: my block is %s, found destination %s", startBlock.getBlockPosition(), currentBlock.getBlockPosition()));
```

Comment:
Do you really need `String.format`? I think that the logging method can also do some string interpolation:

```suggestion
        logger.debug("Looking for a block: my block is {}, found destination {}", startBlock.getBlockPosition(), currentBlock.getBlockPosition()));
```

Reply:
I didn't know that the logger could do string interpolation this way! Thanks for telling me this, this is very useful!